tools/highlight-files/highlight-files.ts

```diff
@@ -26,20 +27,38 @@ function getBazelActionArguments() {
 }
 
 if (require.main === module) {
-  // The script expects the bazel-bin path as first argument. All remaining arguments will be
-  // considered as markdown input files that need to be transformed.
-  const [bazelBinPath, ...inputFiles] = getBazelActionArguments();
-
-  // Walk through each input file and write transformed markdown output to the specified
-  // Bazel bin directory.
-  inputFiles.forEach(inputPath => {
-    const fileExtension = extname(inputPath).substring(1);
-    // Convert "my-component-example.ts" into "my-component-example-ts.html"
-    const baseOutputPath = inputPath.replace(`.${fileExtension}`, `-${fileExtension}.html`);
-    const outputPath = join(bazelBinPath, baseOutputPath);
-    const htmlOutput = highlightCodeBlock(readFileSync(inputPath, 'utf8'), fileExtension);
-
-    writeFileSync(outputPath, htmlOutput);
-  });
+    // The script expects the output directory as first argument. Second is the name of the
+    // package where this the highlight target is declared. All remaining arguments will be
+    // considered as markdown input files that need to be transformed.
+
+    const [outDir, packageName, ...inputFiles] = getBazelActionArguments();
+
+    // Walk through each input file and write transformed markdown output
+    // to the specified output directory.
+    for (const execPath of inputFiles) {
+        // Compute a relative path from the package to the actual input file.
+        // e.g `src/components-examples/cdk/<..>/example.ts` becomes `cdk/<..>/example.ts`.
+        const basePath = relative(packageName, execPath);
+        const fileExtension = extname(basePath).substring(1);
+        const parsed = regionParser(readFileSync(execPath, 'utf8'), fileExtension);
+        for (const [regionName, regionSnippet] of Object.entries(parsed.regions)) {
+            // Create files for each found region
+            if (!regionName) { continue; }
+            const highlightedRegion = highlightCodeBlock(regionSnippet, fileExtension);
+            // Convert "my-component-example.ts" into "my-component-example_region-ts.html"
+            const regionBaseOutputPath = basePath.replace(`.${fileExtension}`,
+              `_${regionName}-${fileExtension}.html`);
+            const regionOutputPath = join(outDir, regionBaseOutputPath);
+            ensureDirSync(dirname(regionOutputPath));
+            writeFileSync(regionOutputPath, highlightedRegion);
```

Comment:
Did you consider putting this loop a separate function? such as: `_detectAndHighlightRegionBlocks()`.

I realized too that the name of the rule (`highlight-files`) might not be best anymore. We should rename it to something more specific for "docs examples" (surely can be in a follow-up).

Reply:
I agree, `highlight-files` seems like it would do the actual highlighting when in fact its writing new files for each example file